src/snowflake/snowpark/file_operation.py

```diff
@@ -207,3 +207,47 @@ def get(
         # connector raises IndexError when no file is downloaded from python connector.
         except IndexError:
             return []
+
+    def put_stream(
+        self,
+        input_stream: IO[bytes],
+        stage_location: str,
+        dest_filename: str,
+        *,
+        parallel: int = 4,
+        auto_compress: bool = True,
+        source_compression: str = "AUTO_DETECT",
+        overwrite: bool = False,
+    ) -> PutResult:
+        """Uploads local files to the stage via file stream.
+
+        Args:
+            input_stream: The input stream from which the data will be uploaded.
+            stage_locatio: The stage and prefix where you want to upload the files.
+            dest_filenam: The filename used to store the file stream in the stage.
+            parallel: Specifies the number of threads to use for uploading files. The upload process separates batches of data files by size:
+
+                  - Small files (< 64 MB compressed or uncompressed) are staged in parallel as individual files.
+                  - Larger files are automatically split into chunks, staged concurrently, and reassembled in the target stage. A single thread can upload multiple chunks.
+
+                Increasing the number of threads can improve performance when uploading large files.
+                Supported values: Any integer value from 1 (no parallelism) to 99 (use 99 threads for uploading files).
+                Defaults to 4.
+            auto_compress: Specifies whether Snowflake uses gzip to compress files during upload. Defaults to True.
+            source_compression: Specifies the method of compression used on already-compressed files that are being staged.
+                Values can be 'AUTO_DETECT', 'GZIP', 'BZ2', 'BROTLI', 'ZSTD', 'DEFLATE', 'RAW_DEFLATE', 'NONE'. Defaults to "AUTO_DETECT".
+            overwrite: Specifies whether Snowflake will overwrite an existing file with the same name during upload. Defaults to False.
+
+        Returns:
+            An object of :class:`PutResult` which represents the results of an uploaded file.
+        """
+        put_result = self._session._conn.upload_stream(
```

Comment:
This is probably a bit out of the scope for this PR, but `upload_stream` does not really handle options like `compress_data`, `overwrite` well in stored proc. We might need to add stream support for `_upload` and use that https://github.com/snowflakedb/Stored-Proc-Python-Connector/blob/main/src/snowflake/connector/cursor.py#L543

Reply:
can't we use this https://github.com/snowflakedb/Stored-Proc-Python-Connector/blob/1b3d7f8d5e54a6aaa8852eb08d4b6150d777e266/src/snowflake/connector/cursor.py#L581?

Reply:
We might be able to. It's actually interesting that the java side does not provide an `overwrite` check at all. The only option is `compress`.

https://github.com/snowflakedb/thundersnow/blob/master/src/main/java/com/snowflake/snowpark_java/FileOperation.java#L158-L160

It depends on how we want our API to look like. If it is the same as Java then this API should just work (even though it does not handle compression today). If we want to have a uniform API just like `file.put`, then it might be easier to go through `file_transfer_agent` where we have overwrite check etc.

Reply:
As discussed offline, let's fail if it is stored proc for now, until we fix the `upload_stream` in sp connector

Reply:
I think you can just add a check here: if the user is in SP and the user calls upload_stream, raise an NotImplementedError immediately?

Reply:
Yeah I agree with @sfc-gh-jdu 's suggestion.